### PR TITLE
issue-11: Replace bolt with nutsdb.

### PR DESCRIPTION
**Pull request by mehoppe**
To replace boltdb with badgerdb per the request in [issue 11](https://github.com/kubescape/kapprofiler/issues/11). It seems badgerdb has similar functionality. Unfortunately I could not get the tests to run as I am a bit new, but I am interested in this project.

### CODE REVIEW

**Comment by slashben** [comment]
> To replace boltdb with badgerdb per the request in [issue 11](https://github.com/kubescape/kapprofiler/issues/11). It seems badgerdb has similar functionality. Unfortunately I could not get the tests to run as I am a bit new, but I am interested in this project.

Awesome! Thank you very much! I will look into it soon

**Comment by slashben** [comment]
Hey @mehoppe , the main reason swapping *boltdb* to something else is this issue: https://github.com/etcd-io/bbolt/issues/541

In short, the way we would like to use *boltdb* (event logging with periodic cleanups) causes it to never really release underlying memory pages to the OS thus leaving us with a leaking application 😞 

Can you please check if *badger* implementation is based on memory mapped files, or not? If it is not, then I have much more confidence regarding if your PR solved the problem.

Thank you!

**Comment by mehoppe** [comment]
Hello @slashben ,

After reviewing badger docs, I may have discovered a way to reduce the use of memory mapping. In v2 of badgerDB, there are 2 options that can be set to force file IO instead of memory mapping for tables and logs.

From the [v2 Readme](https://pkg.go.dev/github.com/dgraph-io/badger/v2#section-readme)
```
Memory Usage
Badger's memory usage can be managed by tweaking several options available in the Options struct that is passed in when opening the database using DB.Open.

-     Options.ValueLogLoadingMode can be set to options.FileIO (instead of the default options.MemoryMap) to avoid memory-mapping log files. This can be useful in environments with low RAM.
-     Mode in which LSM tree is loaded (Options.TableLoadingMode)

```

1. [TableLoadingMode](https://pkg.go.dev/github.com/dgraph-io/badger/v2#Options.WithTableLoadingMode)
2. [LogLoadingMode](https://pkg.go.dev/github.com/dgraph-io/badger/v2#Options.WithValueLogLoadingMode)

Hope that helps!
Mike

**Comment by slashben** [comment]
> 1. [TableLoadingMode](https://pkg.go.dev/github.com/dgraph-io/badger/v2#Options.WithTableLoadingMode)

Sounds like an answer, I will test out and if works I will merge!

File: pkg/eventsink/es.go [hunk 1]
```diff
@@ -101,7 +106,7 @@ func (es *EventSink) Stop() error {
 func (es *EventSink) networkEventWorker() error {
 	for event := range es.networkEventChannel {
 		bucket := fmt.Sprintf("network-%s-%s-%s", event.Namespace, event.PodName, event.ContainerID)
-		err := es.fileDB.Update(func(tx *bolt.Tx) error {
+		err := es.fileDB.Update(func(tx *badger.Tx) error {
```
Comment (slashben) [inline pkg/eventsink/es.go:109]: These lines result with error in compilation
```
undefined: badger.Tx
```

Can you please run `make build` in the root directory?

Comment (mehoppe) [reply pkg/eventsink/es.go]: I have figured out my compilation issues and was able to build this successfully.

**Comment by mehoppe** [comment]
Hello @slashben ,

After finally figuring out my local environment issues, I was able to compile it successfully. In doing so, I was able to more easily discover errors in my ways. I also now understand the use of buckets was a valuable feature, though I think badger could likely accomplish this with prefixes it would have been more effort.  So, I found another solution using the nutsdb package. Nutsdb has favorable performance reviews (from randoms on the internet) and allows using buckets. Nutsdb also has a configuration option to use FileIO instead of memory mapping. Using this solution required some re-factoring of the GetxxxEvents functions within es.go.

**Additional Information**
[nutsdb](https://github.com/nutsdb/nutsdb/tree/master)

[Configuration options for FileIO](https://github.com/nutsdb/nutsdb/blob/master/docs/user_guides/options.md)
`RWMode represents the read and write mode. RWMode includes two options: FileIO and MMap. FileIO represents the read and write mode using standard I/O. And MMap represents the read and write mode using mmap.`

Test Output - lots of FAIL, but TestNewEventSink PASSED (?!)
Screenshot 2023-11-13 at 4 28 54 PM


Thank you for your feedback and for reviewing this work,
Mike

**Comment by mehoppe** [comment]
Tests run after resolving merge conflicts:
Screenshot 2023-11-13 at 8 59 29 PM

**Comment by slashben** [comment]
> Tests run after resolving merge conflicts

I will test this today

**Comment by slashben** [comment]
> Tests run after resolving merge conflicts

On my manual tests it seems to be working, I will try to check memory/CPU consumption also, if it is good I will merge 🚀

**Comment by slashben** [comment]
> Tests run after resolving merge conflicts

On my manual tests it seems to be working, I will try to check memory/CPU consumption also, if it is good I will merge 🚀

**Review by slashben — APPROVED** [review]
Tested
1. Unit tests pass
2. Manual tests in cluster

**Comment by mehoppe** [comment]
Awesome news!

**Pull request merged**